Initial support for viomi.waterheater.e1 devices

Zuz666 opened the pull request:
Just added basic support for 'Viomi Electric Water Heater' devices (viomi.waterheater.e1*): power on/off + status (retrieve properties). Fully tested on real device, works fine.

rytilahti commented on miio/waterheater.py:
```diff
@@ -0,0 +1,133 @@
+class WaterHeaterStatus:
+
+    def __init__(self, data: Dict[str, Any]) -> None:
+        self.data = data
```
Please add an example response showing the data to docstring, here's an example: https://github.com/rytilahti/python-miio/blob/master/miio/heater.py#L56

rytilahti commented on miio/waterheater.py:
```diff
@@ -0,0 +1,133 @@
+        self.data = data
+
+    @property
+    def washStatus(self) -> str:
```
```suggestion
    def status(self) -> str:
```
Which statuses are known? Could it be exposed as an enum?

rytilahti commented on miio/waterheater.py:
```diff
@@ -0,0 +1,133 @@
+        return self.data["velocity"]
+
+    @property
+    def waterTemp(self) -> str:
```
```suggestion
    def water_temperature(self) -> str:
```
Is this a string or integer or float or something else?

rytilahti commented on miio/waterheater.py:
```diff
@@ -0,0 +1,133 @@
+        return self.data["waterTemp"]
+
+    @property
+    def targetTemp(self) -> str:
```
```suggestion
    def target_temperature(self) -> str:
```
Same as above.

rytilahti commented on miio/waterheater.py:
```diff
@@ -0,0 +1,133 @@
+        return self.data["targetTemp"]
+
+    @property
+    def errStatus(self) -> str:
```
```suggestion
    def error(self) -> str:
```
Are error codes/values known?

rytilahti reviewed (changes requested):
Thanks for the PR! A brief initial review: please add an example response, follow the python naming conventions (`foo_bar` instead of camelcased `fooBar`). The property names should be descriptive instead of using the naming used by the protocol itself.

Zuz666 commented:
@rytilahti Thanks for great code review! I will make the code better as you request.

Zorgino commented:
@Zuz666 , any progress on this pr? 🙂

Zuz666 commented:
> @Zuz666 , any progress on this pr? 🙂

@rytilahti, yep, just check the latest commits!

rytilahti commented on miio/waterheater.py:
```diff
@@ -0,0 +1,219 @@
+        properties = SUPPORTED_MODELS.get(
+            self.model, SUPPORTED_MODELS[MODEL_WATERHEATER_E1]
+        )["available_properties"]
+
+        # A single request is limited to 16 properties. Therefore the
+        # properties are divided into multiple requests
+        _props_per_request = 15
+
+        # The E1 is limited to a single property per request
+        if self.model in [MODEL_WATERHEATER_E1]:
+            _props_per_request = 1
+
+        values = self.get_properties(properties, max_properties=_props_per_request)
```
```suggestion
        values = self.get_properties(properties, max_properties=1)
```
I would just simplify this. If there will be another waterheater based on the same series that supports multi-queries, we can adapt this afterwards.

rytilahti commented on miio/waterheater.py:
```diff
@@ -0,0 +1,219 @@
+    def mode(self) -> OperationMode:
+        return OperationMode(self.data["modeType"])
+
+    @property
+    def booking_time_start(self) -> int:
+        return self.data["appointStart"]
+
+    @property
+    def booking_time_end(self) -> int:
+        return self.data["appointEnd"]
```
What is the meaning of these? What do the values (7 and 0) in the example response mean?

rytilahti commented on miio/waterheater.py:
```diff
@@ -0,0 +1,219 @@
+        bacteriostatic_temp: int
+        bacteriostatic_mode = SUPPORTED_MODELS.get(
+            self.model, SUPPORTED_MODELS[MODEL_WATERHEATER_E1]
+        )["bacteriostatic_mode"]
+        if not bacteriostatic_mode:
+            raise WaterHeaterException("Bacteriostatic mode is not supported.")
+        bacteriostatic_temp = SUPPORTED_MODELS.get(
+            self.model, SUPPORTED_MODELS[MODEL_WATERHEATER_E1]
+        )["bacteriostatic_temperature"]
+        mode = OperationMode(self.send("get_prop", ["modeType"])[0])
+        if mode == OperationMode.Termostatic:
+            raise WaterHeaterException("Bacteriostatic operational mode is not supported in Termostatic mode.")
+        return self.send("set_temp", [bacteriostatic_temp])
```
```suggestion
        )["bacteriostatic_mode"]
        
        if not bacteriostatic_mode:
            raise WaterHeaterException("Bacteriostatic mode is not supported.")
            
        bacteriostatic_temp = SUPPORTED_MODELS.get(
            self.model, SUPPORTED_MODELS[MODEL_WATERHEATER_E1]
        )["bacteriostatic_temperature"]
        
        
        mode = OperationMode(self.send("get_prop", ["modeType"])[0])
        if mode == OperationMode.Termostatic:
            raise WaterHeaterException("Bacteriostatic operational mode is not supported in Termostatic mode.")
        return self.send("set_temp", [bacteriostatic_temp])
```
Please add some newlines to separate logical code blocks for easier readability, also on `set_appoint` and `set_mode` :-)

rytilahti commented on miio/waterheater.py:
```diff
@@ -0,0 +1,219 @@
+
+    @property
+    def velocity(self) -> int:
+        return self.data["velocity"]
```
Please add missing docstrings to all publicly exposed API.

rytilahti commented on miio/waterheater.py:
```diff
@@ -0,0 +1,219 @@
+        if mode == OperationMode.Booking:
+            appoint_start = self.send("get_prop", ["appointStart"])[0]
+            appoint_end = self.send("get_prop", ["appointEnd"])[0]
+#            raise WaterHeaterException("Booking operational mode is not supported in set_mode use set_appoint instead of.")
```
Should this be removed?

Zuz666 replied:
Yes! Just a stub when the feature was not implemented.

rytilahti reviewed (commented):
Thanks for the update! I added a couple of more comments. Could you also do the following:
* Add an entry to README.md 
* Move the code to be under `miio/integrations/heater/viomiwaterheater/` - there are other heaters so we may have a common API among all of them at some point :-)

Zuz666 commented:
> * Add an entry to README.md
> * Move the code to be under `miio/integrations/heater/viomiwaterheater/` - there are other heaters so we may have a common API among all of them at some point :-)

@rytilahti Job done! Please perform a code review.

rytilahti commented on miio/device.py:
```diff
@@ -249,10 +249,10 @@ def test_properties(self, properties):
 
         def ok(x):
-            click.echo(click.style(x, fg="green", bold=True))
+            click.echo(click.style(str(x), fg="green", bold=True))
```
Are the changes to this file necessary? If yes, could you please open a separate PR as it's not related to the water heater itself?

Zuz666 replied:
Done https://github.com/rytilahti/python-miio/pull/1180

rytilahti commented on miio/integrations/waterheater/viomi/viomiwaterheater.py:
```diff
@@ -0,0 +1,271 @@
+
+    @property
+    def cleaning_required(self) -> int:
+        """Not 0 when cleaning the device is required."""
+        return self.data["needClean"]
```
Please convert the return value to bool:
```
return self.data["needClean"] != 0
```
(or `> 0`?)

rytilahti commented on miio/integrations/waterheater/viomi/viomiwaterheater.py:
```diff
@@ -0,0 +1,271 @@
+                "Bacteriostatic operational mode is "
+                "not supported in Thermostatic mode."
+            )
+        else:
+            return self.send("set_temp", [bacteriostatic_temp])
```
```suggestion

        return self.send("set_temp", [bacteriostatic_temp])
```
No need for else here.

rytilahti commented on miio/integrations/waterheater/viomi/viomiwaterheater.py:
```diff
@@ -0,0 +1,271 @@
+            lambda booking_time_start, booking_time_end: "Setting up the Booking mode operational interval from: %s "
+            % booking_time_start
+            + "to: %s " % booking_time_end
+            + "(duration: %s)"
+            % (
+                booking_time_end - booking_time_start
+                if booking_time_end - booking_time_start > 0
+                else booking_time_end - booking_time_start + 24
+            )
```
Maybe remove the duration from the output, as it makes reading this rather complicated?

Zuz666 replied:
@rytilahti, цhy? After substituting the parameters, it looks like this:
_Setting up the Booking mode operational interval from: 6 to: 0 (duration: 18 hours)._

This allows us to better understand whether the interval was set correctly.
Alternatively, we can change the parameters this command and set the start time and duration.

rytilahti replied:
Okay, that's fine by me :+1:

rytilahti commented on miio/integrations/waterheater/viomi/viomiwaterheater.py:
```diff
@@ -0,0 +1,271 @@
+            )
+        ),
+    )
+    def set_appoint(self, booking_time_start, booking_time_end):
```
I'd say that it's worth to consolidate the verbiage, either using booking or appoint(ment) for both variables and for the method name? As I don't know the details, I'm not sure which one is better though.

rytilahti commented on miio/integrations/waterheater/viomi/viomiwaterheater.py:
```diff
@@ -0,0 +1,271 @@
+                "Booking mode operational interval parameters "
+                "must be within [0, 23]."
+            )
+        else:
+            return self.send("set_appoint", [1, booking_time_start, booking_time_end])
```
```suggestion
        return self.send("set_appoint", [1, booking_time_start, booking_time_end])
```
Please remove also other unnecessary elses. The error conditions should be checked as early as possible, and there is no need to perform any checks for the call itself when the preconditions are already checked.

rytilahti reviewed (commented):
Looking good, I added a couple of very minor changes.

Could you also please add at least tests to the container functionality (see https://github.com/rytilahti/python-miio/pull/1174#pullrequestreview-799144026)?

Zuz666 commented:
> Looking good, I added a couple of very minor changes.
> 
> Could you also please add at least tests to the container functionality (see [#1174 (review)](https://github.com/rytilahti/python-miio/pull/1174#pullrequestreview-799144026))?

Some code for testing done but how can I run this test? Newbie to python. )))

rytilahti commented:
Simply execute `pytest` and it should pick it automatically (as long as the filename and the test functions follow the `test_*` pattern) :-)

Zuz666 commented:
Hi @rytilahti,

I have fully rebased this PR on top of the current `master` and modernized the codebase to align with the current architecture!

Here is a summary of the updates:
* **Rebase & Modernization:** The code was moved to `miio.integrations.viomi.waterheater` following the latest manufacturer/device folder structure.
* **Decorators & API:** Transformed `ViomiWaterHeaterStatus` to inherit from `DeviceStatus` and use `@sensor` and `@setting` decorators. Dropped obsolete patches and cleaned up redundant `else` blocks in control methods.
* **Unit Tests:** Added full unit tests (`test_viomiwaterheater.py`) utilizing `DummyDevice`. All tests successfully pass and validate status parsing and operations.
* **Real Device Verification:** Executed a live smoke test against a real `viomi.waterheater.e1` device, verifying proper parsing and typing of the properties.

Please take a look and let me know if anything else is needed!

Zuz666 commented:
Pushed a fix for the CI linting issues (pre-commit hooks: docstring placement, zip strict parameter, and f-strings).

Zuz666 commented:
Added unit tests for `set_bacteriostatic_mode` to achieve 100% patch coverage and resolve the Codecov warning.

Zuz666 commented:
Also added a computed `bacteriostatic_mode_active` sensor to `ViomiWaterHeaterStatus`. Since the device relies purely on `targetTemp == 80` for this feature and has no native toggle, this computed property will give downstream integrations (like Home Assistant) a convenient binary sensor to observe the mode's status.

Zuz666 commented:
Hi @rytilahti, all prior review comments have been addressed after the rebase. All 14 CI checks are passing and all review threads are resolved.

The PR is technically clean; the remaining `CHANGES_REQUESTED` state appears to be the review from 2020, before the current rebase. Could you please re-review it, or dismiss the obsolete review state, when you have a chance?

Thanks!
